src/ray/core_worker/core_worker_shutdown_executor.cc

```diff
@@ -32,25 +32,26 @@ void CoreWorkerShutdownExecutor::ExecuteGracefulShutdown(
     std::string_view exit_type,
     std::string_view detail,
     std::chrono::milliseconds timeout_ms) {
-  RAY_LOG(INFO) << "Executing graceful shutdown: " << exit_type << " - " << detail
-                << " (timeout: " << timeout_ms.count() << "ms)";
-
-  // For actors, perform cleanup before shutdown proceeds.
-  if (!core_worker_->worker_context_->GetCurrentActorID().IsNil() &&
-      core_worker_->actor_shutdown_callback_) {
-    RAY_LOG(INFO) << "Calling actor shutdown callback before shutdown";
-    core_worker_->actor_shutdown_callback_();
-  }
+  RAY_LOG(DEBUG) << "Executing graceful shutdown: " << exit_type << " - " << detail
+                 << " (timeout: " << timeout_ms.count() << "ms)";
 
   if (core_worker_->options_.worker_type == WorkerType::WORKER) {
-    // Running in a main thread.
-    // Asyncio coroutines could still run after CoreWorker is removed because it is
-    // running in a different thread. This can cause segfault because coroutines try to
-    // access CoreWorker methods that are already garbage collected. We should complete
-    // all coroutines before shutting down in order to prevent this.
-    if (core_worker_->worker_context_->CurrentActorIsAsync()) {
+    // For asyncio actors, terminate the asyncio thread early to prevent coroutines
+    // from accessing torn-down C++ state during shutdown.
+    if (!core_worker_->worker_context_->GetCurrentActorID().IsNil() &&
+        core_worker_->worker_context_->CurrentActorIsAsync()) {
+      RAY_LOG(DEBUG) << "Terminating asyncio thread early for actor shutdown";
       core_worker_->options_.terminate_asyncio_thread();
     }
+
+    // For actors, perform Python-side cleanup before shutdown proceeds.
+    if (!core_worker_->worker_context_->GetCurrentActorID().IsNil() &&
+        core_worker_->actor_shutdown_callback_) {
+      RAY_LOG(DEBUG) << "Calling actor shutdown callback";
+      core_worker_->actor_shutdown_callback_();
```

Comment:
seems like terminating the asyncio thread should just be part of the actor shutdown callback? it feels like a bit of an abstraction leak at the moment